### DISCUSSION TITLE
Can't run lua script

### DISCUSSION THREAD

**Discussion by GuillaumeDehaene** [quarto-dev/quarto-cli#9057] [Q&A] [answered]
### Description

Hello everyone

I'm trying to get back into developping a lua filter to simplify how math can be inputted in pandoc and quarto Markdown.
As part of that, I would like to have:
- a list of examples pairing raw markdown with their conversion
- a test script to check that these examples are correctly converted

However, as part of that, I can't seem to be able to use `quarto run test.lua`
Am I doing something wrong? Here is the error trace.

```
PS D:\Desktop\Dojo\quarto\filters\simple_tex_math> quarto run test.lua
Error running filter C:\Users\Guillaume\AppData\Local\Programs\Quarto\share\filters\init\init.lua:
...ppData\Local\Programs\Quarto\share\filters\init\init.lua:4: module 'readqmd' not found:
        no field package.preload['readqmd']
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\lua\readqmd.lua'
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\lua\readqmd\init.lua'
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\readqmd.lua'
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\readqmd\init.lua'
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\..\share\lua\5.4\readqmd.lua'        
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\..\share\lua\5.4\readqmd\init.lua'   
        no file '.\readqmd.lua'
        no file '.\readqmd\init.lua'
        no file 'D:\Lua\lua\readqmd.luac'
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\readqmd.dll'
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\..\lib\lua\5.4\readqmd.dll'
        no file 'C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin\tools\loadall.dll'
        no file '.\readqmd.dll'
stack traceback:
        ...ppData\Local\Programs\Quarto\share\filters\init\init.lua:4: in main chunk
PS D:\Desktop\Dojo\quarto\filters\simple_tex_math>
```

Other debug information:
```
- IDE VSCode 1.87.1
- OS  Windows 11 version 23H2

>> quarto check
Quarto 1.5.24
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.5.24
      Path: C:\Users\Guillaume\AppData\Local\Programs\Quarto\bin
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[>] Checking LaTeX....................OK
      Using: Installation From Path
      Path: C:\texlive\2023\bin\windows
      Version: 2023

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....(None)

      Unable to locate an installed version of Python 3.
      Install Python 3 from https://www.python.org/downloads/

[>] Checking R installation...........(None)

      Unable to locate an installed version of R.
      Install R from https://cloud.r-project.org/
```

The issue was also present on my earlier quarto 1.4.X and 1.3.X installs. I've also tried to uninstall and reinstall.
Quarto filters run perfectly and no other issue with lua is apparent

**Comment by cderv** [comment]
Just out of curiosiy, what do you do in `test.lua` ? 

I think I can reproduce different issues with `quarto run` used on Lua script. So something is not ok - it should work per our doc: https://quarto.org/docs/projects/scripts.html#periodic-scripts

````
quarto run test.lua
Error running filter C:\Users\chris\scoop\apps\quarto-prerelease\current\share\filters\init\init.lua:
cannot open C:\Users\chris\scoop\apps\quarto-prerelease\current\share\filters\init\init.lua: No such file or directory
````

I don't even have the `init.lua` file when this is installed. 

Thanks for the report.

**Reply by cderv** [reply, marked as answer]
I opened 
* https://github.com/quarto-dev/quarto-cli/issues/9059

to track this. 

> I don't even have the init.lua file when this is install

@GuillaumeDehaene I am even surprised that you get the file found in `Quarto 1.5.24` as we don't have this file at this path anymore. 

Like you still got some old file in your install at `C:\Users\Guillaume\AppData\Local\Programs\Quarto` 🤔 

I am mentioning this because probably if the file you have is old, then it won't call the right dependency hence the issue you see.

**Reply by GuillaumeDehaene** [reply]
At the time of writing, test.lua was just print("hello world") so the code inside was definitely not the issue

In the final version, I would like to make sure that my filter has a test suite:
- because sharing untested code is bad
- because I've been already been burned quite a bit during the development of my math parser, and I need it to avoid regressions during my future dev

I have a clownish workaround:
- I wrote a python script to convert example.qmd to markdown (with `quarto run example.qmd --to markdown`
- I check that the resulting markdown is what I expect

If you have a better idea, that would be wonderful!

Edit: Indeed, the init.lua file dates from 24/10/23 so it's probaly from version 1.3.X
I'll make a not to uninstall and purge everything in the future.

**Reply by cderv** [reply]
> (with quarto run example.qmd --to markdown

You mean `quarto render` right ? 

> I'll make a not to uninstall and purge everything in the future.

I would think that installing the MSI should have removed and replaced... If it seems it does not do it, please do tell us. 

> I would like to make sure that my filter has a test suite:

You can probably indeed test your lua function using `quarto run` is you use some testing in lua itself. 

Otherwise you can also leverage adding the filter in YAML of a `.qmd` and testing the output. You can even throw error within Lua to stop the rendering is some AST processing is not respected. It all depend on what your Lua. 

The recent change in 1.4 allows to add some filters at specific place of the quarto rendering (https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases) you can also leverage that to add some filters to test output and processing. 

Overall, this is not yet easy to test extensions and filter, but we'll come up in a solution at some point I hope.

**Reply by GuillaumeDehaene** [reply]
1. Yes, quarto render.
2. I see that you've already merged the change. Can you please tell me how long I should wait before the merge is reflected in the pre-release dowload at https://quarto.org/docs/download/ ?
3. I'll report once the pre-release reflects the changes whether init.lua is still there or not

**Reply by mcanouil** [reply]
It varies, a pre-release build is not triggered at every changes in `main`.
You can subscribe to the releases on GitHub and wait for 1.5.25.
(https://github.com/quarto-dev/quarto-cli/releases)

<img width="325" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/520985dd-0f5b-439f-8238-ef1f8ee90045">

**Reply by GuillaumeDehaene** [reply]
Hello,

I've installed 1.5.25 and the fix works perfectly:
- installing did not remove the old init.lua file
- regardless, `quarto run test.lua` now works
- uninstalling quarto removes a bunch of files from AppData/Local/Programs/Quarto but somehow misses init.lua
- files can be manually erased

Thank you so much!

Best
Guillaume

**Reply by cderv** [reply]
> installing did not remove the old init.lua file

Thanks for the feedback. I'll look into this - or pay attention if anyone reports it again. But 1.5 will fix this anyway. 

Thank a lot for the report! This was a very old bug I am happy we fixed!
